### PR TITLE
Define algorithm input and output types.

### PR DESCRIPTION
**Pull request by jyasskin**
Both the transform, hash, and proof algorithms that cryptographic suites have to define, and this
spec's Verify Proof interface.

This removes the underspecified "proof options" field in favor of saying they're members of the `proof` object for verification, or defined by the input cryptographic suite for signing.

I started this on my way to trying to have "verify" also return the verified document", but I decided this was a big enough change to start with.

Apologies for any obvious mistakes I've made here.

***
<a href="https://pr-preview.s3.amazonaws.com/jyasskin/vc-data-integrity/pull/226.html" title="Last updated on Jan 2, 2024, 7:58 PM UTC (e7a445f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/226/3140804...jyasskin:e7a445f.html" title="Last updated on Jan 2, 2024, 7:58 PM UTC (e7a445f)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -2440,6 +2440,33 @@ <h2>Cryptographic Suites</h2>
 verification of the data.
         </li>
         <li>
+          The specification MUST define a <dfn>suite selection algorithm</dfn> that accepts a [=JSON
+          object=] |proof| and uses the properties described in [[[#proofs]]] to return either
+          failure or a [=VC cryptographic suite=] [[Infra]] [=struct=]. This algorithm SHOULD be
```
Comment (jyasskin) [inline index.html:2447]: This might be able to take just the `type` and `cryptosuite`, but I wasn't sure if y'all wanted to support some suites that also need other properties to figure out the algorithms.

File: index.html [hunk 2]
```diff
@@ -2863,45 +2890,28 @@ <h3>Add Proof</h3>
...
-<dfn>unsecured data document</dfn>. Required inputs are an
-<a>unsecured data document</a> (<var>unsecuredDocument</var>) and
-<a>proof options</a> (<var>options</var>). The
-<a>proof options</a> MUST contain a type identifier for the
-<a>cryptographic suite</a> (<var>type</var>) and any other properties needed by
-the <a>cryptographic suite</a> type; an identifier for the
-<a>verification method</a> (<var>verificationMethod</var>) that can be used to
-verify the authenticity of the proof; an [[!XMLSCHEMA11-2]] `dateTimeStamp`
-string (<var>created</var>) containing the current date and time,
-accurate to at least one second, in Universal Time Code format. A
```
Comment (jyasskin) [inline index.html:2877]: I dropped the timestamp because the signing algorithm can retrieve the current date and time just as well as the caller.

File: index.html [hunk 3]
```diff
@@ -2863,45 +2890,28 @@ <h3>Add Proof</h3>
...
 The following algorithm specifies how to add a digital proof to a document,
 which  can be used to verify the authenticity and integrity of an
-<dfn>unsecured data document</dfn>. Required inputs are an
-<a>unsecured data document</a> (<var>unsecuredDocument</var>) and
-<a>proof options</a> (<var>options</var>). The
-<a>proof options</a> MUST contain a type identifier for the
-<a>cryptographic suite</a> (<var>type</var>) and any other properties needed by
-the <a>cryptographic suite</a> type; an identifier for the
-<a>verification method</a> (<var>verificationMethod</var>) that can be used to
-verify the authenticity of the proof; an [[!XMLSCHEMA11-2]] `dateTimeStamp`
```
Comment (jyasskin) [inline index.html:2875]: The verification method is being used to smuggle in a private key, but that's far from obvious in the old interface. The new interface also doesn't say it explicitly, but the private key should go into the _proof creation options_ that the suite specifies.

File: index.html [hunk 4]
```diff
@@ -2440,6 +2440,33 @@ <h2>Cryptographic Suites</h2>
...
+          registered in [[[VC-SPECS]]]. A
+          <dfn class="export">VC cryptographic suite</dfn> [=struct=] has the following
+          [=struct/items=].
+          <dl data-dfn-for="cryptographic suite">
+            <dt><dfn>transform</dfn></dt>
+            <dd>A [=transformation algorithm=] that takes a [=JSON object=] as input and returns a
+            [=JSON object=].</dd>
+            <dt><dfn>hash</dfn></dt>
+            <dd>A [=hashing algorithm=] that takes a [=JSON object=] as input and returns a [=byte
+            sequence=].</dd>
```
Comment (jyasskin) [inline index.html:2457]: There's a note that the hashing algorithm can actually return things that aren't just single byte sequences. The suite could specify a type for this, or we could just say it returns an instance of a type that can be passed to the **serialize proof** algorithm, or we could merge hashing with serializing.

File: index.html [hunk 5]
```diff
@@ -3000,21 +3013,31 @@ <h3>Verify Proof</h3>
...
-<var>challenge</var>. The <var>expectedProofPurpose</var> value is used to
-ensure that the <var>proof</var> was generated by the proof creator for the
-expected reason by the verifier, such as "authentication". The <var>domain</var>
-value is used by the proof creator to lock a proof to a particular security
-domain, and used by the verifier to ensure that a proof is not being used across
-different security domains. The <var>challenge</var> value is used by the
-verifier to ensure that an attacker is not replaying previously created proofs.
-Additional options are expected to be provided by libraries to ensure that
-digital proof time stamps do not deviate more than an acceptable time frame
-for a given use case. A <dfn>verification result</dfn> is produced as output.
```
Comment (jyasskin) [inline index.html:3035]: I dropped this for now, but from some of the discussion elsewhere, it sounded like the WG wants these time-checking operations to be part of "validation" rather than "verification". If that's the case, this algorithm should *return* some extra data for the validator to act on.

File: index.html [hunk 6]
```diff
@@ -3000,21 +3013,31 @@ <h3>Verify Proof</h3>
...
+<var>proof</var> was generated by the proof creator for the expected reason by the verifier, such as
+"authentication"
+          </li>
+          <li>
+|domain|, an optional [=string=], used by the proof creator to lock a proof to a particular security
+domain, and used by the verifier to ensure that a proof is not being used across different security
+domains
+          </li>
+          <li>
+|challenge|, an optional [=string=] [=challenge=], used by the verifier to ensure that an attacker
```
Comment (jyasskin) [inline index.html:3059]: Are these actually strings?

File: index.html [hunk 7]
```diff
@@ -3046,41 +3074,30 @@ <h3>Verify Proof</h3>
...
-Let <var>isProofVerified</var> be the result of running the
-<a>proof verification algorithm</a> associated with the
-<a>cryptographic suite</a> with the <var>hashData</var> and <var>options</var>
-parameters provided as inputs to the algorithm.
+Let |isProofVerified:valid or invalid| be the result of running |suite|'s
+[=cryptographic suite/verify proof=] algorithm on <var>hashData</var> and <var>proof</var>.
           </li>
           <li>
-If <var>options</var>.<var>domain</var> is set and it does not match
+If <var>domain</var> was given, and it does not match
```
Comment (jyasskin) [inline index.html:3114]: "match" is a yellow flag in my spec reviews. Usually it should be "equal", but sometimes people mean a different comparison algorithm that needs to be specified. That fix should go in a different PR.

File: index.html (hunk 3 above)
Comment (dlongley) [reply index.html:2875]: An important thing we want to avoid is saying that any algorithms require private key material to be passed in opposed to an interface for performing signing. It's often the case that signals are sent to implementers that they ought to require people to pass in private key material when requiring a signer interface instead enables much better security practices (such as HSM use, etc.). These signals also sometimes create headaches where implementers choose to provide synchronous interfaces (as opposed to asynchronous ones) which do not even allow people to "squeeze in" HSM-backed signing mechanisms.

Comment (jyasskin) [reply index.html:2875]: Good point. I've tried to incorporate this in 3d6aa172a485ade9d5a5167d2a7b7ff4221064ad.

**Review by msporny — APPROVED** [review]
Approving so we can get these changes into the specification and rework the language. 

There are now merge conflicts with @Wind4Greg's previous PR that got merged. I tried for an hour or so to resolve the differences and was unsure of the result.

@jyasskin, can you please rebase and fix the conflicts so we can do a clean merge?

**Comment by jyasskin** [comment]
@msporny Thanks for trying. I've rebased, and I think it was just a textual conflict and not a problem with our meanings. It should be good to go.

File: index.html (hunk 7 above)
Comment (msporny) [reply index.html:3114]: The intention here was "equal" ... as in string equality. Can fix in a future iteration as an editorial fix.

**Comment by msporny** [comment]
@jyasskin wrote:
> I've rebased, and I think it was just a textual conflict and not a problem with our meanings. It should be good to go.

Great, thank you! ... and thank you for this PR, it's a solid improvement to the interfaces between VCDM, the VC Specs Directory, and the various Data Integrity specs... I'll try to follow it up w/ interface PRs to each of the Data Integrity specs over the coming weeks.

My focus over the next month will be getting VCDM v2.0 and vc-bitstring-status-list into CR... I'll then circle back around to the Data Integrity specs.

File: index.html [hunk 8]
```diff
@@ -3064,41 +3099,30 @@ <h3>Verify Proof</h3>
...
           <li>
-Let <var>transformedData</var> be the result of
-<a href="#dfn-transformation">transforming</a> the <var>unsecuredDocument</var>
-according to a <a>transformation algorithm</a> associated with the
-<a>cryptographic suite</a> specified in <var>proof</var> and the
-<var>options</var> parameters provided as inputs to the algorithm.
-The type of <a>cryptographic suite</a> is specified by the
-<var>proof.type</var> value and MAY be further described by
-<a>cryptographic suite</a>-specific properties expressed in <var>proof</var>.
+Let <var>transformedData</var> be the result of running |suite|'s'
```
Comment (TallTed) [inline index.html:3102]:
```suggestion
Let <var>transformedData</var> be the result of running |suite|'s
```

File: index.html [hunk 9]
```diff
@@ -3064,41 +3099,30 @@ <h3>Verify Proof</h3>
...
+[=cryptographic suite/transform=] algorithm on the <var>unsecuredDocument</var>.
         </li>
           <li>
-Let <var>hashData</var> be the result of
-<a href="#dfn-hashing">hashing</a> the <var>transformedData</var>
-according to a <a>hashing algorithm</a> associated with the
-<a>cryptographic suite</a> specified in the <var>proof</var> and
-<var>options</var> parameters provided as inputs to the algorithm.
+Let <var>hashData</var> be the result of running |suite|'s'
```
Comment (TallTed) [inline index.html:3106]:
```suggestion
Let <var>hashData</var> be the result of running |suite|'s
```

File: index.html [hunk 10]
```diff
@@ -3064,41 +3099,30 @@ <h3>Verify Proof</h3>
...
+[=cryptographic suite/hash=] algorithm on the <var>transformedData</var>.
           </li>
           <li>
-Let <var>isProofVerified</var> be the result of running the
-<a>proof verification algorithm</a> associated with the
-<a>cryptographic suite</a> with the <var>hashData</var> and <var>options</var>
-parameters provided as inputs to the algorithm.
+Let |isProofVerified:valid or invalid| be the result of running |suite|'s
```
Comment (TallTed) [inline index.html:3110]:
```suggestion
Let |isProofVerified:valid| or |isProofVerified:invalid| be the result of running |suite|'s
```

File: index.html (hunk 5 above)
Comment (msporny) [reply index.html:3035]: We wanted the ability to specify the time for which a proof was being created, or verified. There are use cases where you want to verify against a time in the past, or sign for a time in the future. I believe this line was there to ensure that one could have that flexibility. We can add this back in a future PR.

**Review by TallTed — CHANGES_REQUESTED** [review]

File: index.html (hunk 1 above)
Comment (msporny) [reply index.html:2447]: Yes, we wanted to give implementations the latitude to accept other parameters that they might need. The algorithms here are supposed to define the bare minimum necessary, but not restrict implementations for doing more than the bare minimum.

File: index.html (hunk 6 above)
Comment (msporny) [reply index.html:3059]: Yes, they are.

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

File: index.html (hunk 10 above)
Comment (msporny) [reply index.html:3110]: Ah, crap... I merged w/o seeing these... raise another PR to fix these, please, @TallTed.... note that I'm getting ready to merge the other PR from @jyasskin in a bit, so maybe wait a day for this stuff to settle before doing your PR.

Comment (jyasskin) [reply index.html:3110]: @TallTed's suggestion isn't the right fix here, although I agree that what I wound up with is awkward. This is the `|variable:type|` syntax from https://respec.org/docs/#giving-variables-data-types, and the type of the variable is actually the enum type of "valid" or "invalid". Respec doesn't accept `"` inside its |var| syntax, and it doesn't appear to have another way of attaching types, so I dropped the quotes. One could argue that I should have instead given up on assigning a type to this variable.

Comment (TallTed) [reply index.html:3110]: @jyasskin — I think you're in a better position than I to log a ReSpec issue, about how to type such enum-typed vars ... whether that leads to some complex manual process, an improvement of ReSpec, or otherwise.

Comment (jyasskin) [reply index.html:3110]: Indeed, I am. And in figuring out what issue to file, I found the markup I should have used: https://github.com/w3c/respec-web-services/pull/399 and https://github.com/w3c/respec/wiki/data-type.

Comment (TallTed) [reply index.html:3110]: @jyasskin — As you've noted that _Respec doesn't accept `"` inside its `|var|` syntax_, it seems that some addition detail should be in [that wiki](https://github.com/w3c/respec/wiki/data-type) (and hence the proper ReSpec documentation) to cover such scenarios as this enum typing.

Comment (jyasskin) [reply index.html:3110]: Sure, done: https://github.com/w3c/respec/wiki/data-type. Along with the bug to make it support quotes.
